pkg/promtail/targets/windows/bookmark.go

```diff
@@ -0,0 +1,90 @@
+//+build windows
+
+package windows
+
+import (
+	"io/ioutil"
+	"os"
+
+	"github.com/grafana/loki/pkg/promtail/targets/windows/win_eventlog"
+	"github.com/spf13/afero"
+)
+
+type bookMark struct {
+	handle win_eventlog.EvtHandle
+	file   afero.File
+	isNew  bool
+
+	buf []byte
+}
+
+// newBookMark creates a new windows event bookmark.
+// The bookmark will be saved at the given path. Use save to save the current position for a given event.
+func newBookMark(path string) (*bookMark, error) {
+	// 4kb buffer for rendering bookmark
+	buf := make([]byte, 4<<10)

```

Comment:
4kb will not be enough for complex event filters

Reply:
You have an example ? what would be a good value ?

Reply:
By hitting some filters in Custom View of Windows Event Viewer and exporting resulting filter XML, I am easily getting a 8kb file. Bookmark with that filter takes a little bit more space.

Telegraf uses a 16kb buffer, and it's a better default value, enough for most cases.

The best way would be going full winapi: calling _EvtRender with zero as buf pointer value. It will return error, but also put the required length into bufferUsed. So then we can make buffer of required length and call _EvtRender again, this time with the newly created buffer pointer. It is a less performant way, and there can be a headache of memory allocating and freeing, but it is bulletproof.

Thank you very much for this PR btw! I made like 75% of porting Telegraf code to Loki, but got stuck at the polling loop, didn't find the satisfying method to do polls and got carried away by other duties. I like Loki, but extending it is such nontrivial work comparing to Telegraf, so many core files to change with custom targets, it is very unfriendly.

Reply:
> I like Loki, but extending it is such nontrivial work comparing to Telegraf, so many core files to change with custom targets, it is very unfriendly.

Sorry about that ! We'll try to get better at this.

Reply:
I'll increase the buffer :) before we cut a new release